python/ray/ml/checkpoint.py

```diff
@@ -1,17 +1,90 @@
 import abc
-from typing import Dict, Optional
+import json
+import shutil
+import tarfile
+import tempfile
+
+import cloudpickle as pickle
+import os
+from typing import Any, Type, Optional, Dict, List
 
 import ray
+from ray.util.ml_utils.cloud import (
+    upload_to_bucket,
+    is_cloud_target,
+    download_from_bucket,
+)
+
+
+def write_metadata(path: str, metadata: Dict[str, Any]) -> None:
+    metadata_file = os.path.join(path, ".checkpoint_metadata")
+    with open(metadata_file, "wt") as fp:
+        json.dump(metadata, fp)
+
+
+def load_metadata(path: str) -> Dict[str, Any]:
+    metadata_file = os.path.join(path, ".checkpoint_metadata")
+    if os.path.exists(metadata_file):
+        with open(metadata_file, "rt") as fp:
+            return json.load(fp)
+
+
+def temporary_checkpoint_dir() -> str:
+    return tempfile.mkdtemp(prefix="checkpoint_tmp_", dir=os.getcwd())
+
+
+def _pack(path: str) -> bytes:
+    _, tmpfile = tempfile.mkstemp()
+    with tarfile.open(tmpfile, "w:gz") as tar:
+        tar.add(path, arcname="")
+
+    with open(tmpfile, "rb") as f:
+        stream = f.read()
+
+    os.remove(tmpfile)
+    return stream
+
+
+def _unpack(stream: bytes, path: str) -> str:
+    _, tmpfile = tempfile.mkstemp()
+
+    with open(tmpfile, "wb") as f:
+        f.write(stream)
+
+    with tarfile.open(tmpfile) as tar:
+        tar.extractall(path)
+
+    os.remove(tmpfile)
+    return path
 
 
 class Checkpoint(abc.ABC):
-    """Checkpoint interface"""
+    """Checkpoint interface.
+
+    This implementation provides interfaces to translate between
+    different checkpoint storage locations: Local FS storage, remote
+    node FS storage, data object, and cloud storage location.
+
+    The following metadata keys are introduced for correct serialization
+    and deserialization of checkpoints:
+
+        is_fs_checkpoint (bool): If this is set, this checkpoint was
+            created from a FS location (e.g. a directory) and can thus
+            be deserialized into a FS location again for downstream processing.
+        is_data_checkpoint (bool): If this is set, this checkpoint was
+            created from a data object (e.g. a dictionary) and can
+            thus be deserialized into a data checkpoint again for downstream
+            processing.
+    """
 
     def __init__(self, metadata: Optional[Dict] = None):
         self.metadata = metadata or {}
 
+    def __eq__(self, other):
```

Comment:
For my own understanding, can you comment on how this is going to be used? So that I can imagine what should be implemented here.

Reply:
`__eq__` is just the equality operator (e.g. `checkpoint_1 == checkpoint_2`). checkpoints should be regarded as "the same" if they point to the same location and contain the same data

Reply:
yes, I guess my question is under what circumstances would we check if `checkpoint_1 == checkpoint_2`?

Reply:
We compare checkpoint locations often e.g. in the CheckpointManager (which will use the new checkpoint class in a future PR).

Generally this is a user-facing API, so we should make sure it behaves as users would expect it. And in that case ensuring ``Checkpoint.from_directory(path) == Checkpoint.from_directory(path)`` makes sense.

This is especially relevant as the data objects themselves are private (e.g. `checkpoint._local_path`). Without this class, `checkpoint_path == checkpoint_path` and `checkpoint_dict == checkpoint_dict` will work as expected, so we should make sure this is also the case when using this wrapper class.

Reply:
I feel like this is a strange function. For one, it could be very expensive (byte-wise compare a lot of data).

It would be better to not define this, and allow more specific, well-defined attribute accessors like `.get_local_path()` etc. which can be used for comparisons.